examples/python/celsius_connectors/at_least_once_feed

```diff
@@ -0,0 +1,276 @@
+            break
+    ```
+    """
+    def __init__(self, args=None, required_params=[], optional_params=[]):
+        AtLeastOnceSourceConnector.__init__(self, args, required_params,
+                                            optional_params)
+        self.sources = {} # stream_id: source instance
+        self.keys = []
+        self._idx = -1
+        self.joining = set()
+
+    def stream_updated(self, stream):
+        super(MultiFileReaderConnector, self).stream_updated(stream)
+        source = self.sources.pop(stream.id, None)
+        # stream already in sources
+        if source:
+            # if stream is closed, close the file, then remove the source
+            if not stream.is_open:
+                source.close()
+                # don't put it back in sources...
+            # otherwise... check if you need to reset its position
+            else:
+                # if stream just joined, remove it from joining
+                if stream.id in self.joining:
+                    self.joining.remove(stream.id)
+                    if stream.point_of_ref != source.point_of_ref:
+                        source.reset(stream.point_of_ref)
+                # put it back
+                self.sources[stream.id] = source
+        # stream is new to us
+        else:
+            if stream.is_open:
+                # This is an error for MultiFileReaderConnector
+                raise ConnectorError("Can't open a new source from a stream. "
+                                     "Please use the add_source interface.")
+            # nothing to do with a closed stream for which there is no source
+
+    def add_source(self, source):
+        # add to self.sources
+        _id = self.get_id(source.name)
+        self.sources[_id] = source
+        self.keys.append(_id)
+        # add to joining set so we can control the starting sequence
+        self.joining.add(_id)
+
+        # send a notify
+        self.notify(_id, source.name, source.point_of_ref())
+
+    @staticmethod
+    def get_id(text):
+        """
+        Get a positive U64 from the I64 of the hash of text
+        """
+        return ctypes.c_size_t(hash(text)).value
+
+    # Make this class an iterable:
+    def __next__(self):
+        if len(self.keys) > 0:
+            try:
+                # get next position
+                self._idx = (self._idx + 1) % len(self.keys)
+                # get key of that position
+                key = self.keys[self._idx]
+                # get source at key
+                source = self.sources[key]
+                # get value from source
+                value, point_of_ref = next(source)
+                # send it as a message
+                msg = cwm.Message(
+                    stream_id = key,
+                    flags = cwm.Message.Key,
+                    message_id = point_of_ref,
+                    key = source.key,
+                    message = value)
+                self.write(msg)
```

Comment:
@nisanharamati: For credit flow, I think this call will block until `self.credits > 0` in a naive implementation

Reply:
@strmpnk : Maybe a PID controller is a good way to handle this https://en.wikipedia.org/wiki/PID_controller

Reply:
asyncore provides two useful async handlers here:

2. ```python
    def handle_write(self): 
        self.initiate_send()
        # starts a synchronous send of whatever's in `producer_fifo`
        # But we can override this with, say, pull from user source iterator instead of fifo
        # so instead of buffering in the connector client, buffering is the user's responsibility
        # and this pulls as fast as it can go.
        # optionally, we can pull more than one item before initiating the synchronous send
    ```

2.  ```python
    def writable (self):
        # True if can write (e.g. connection is open and credits > 0)
        "predicate for inclusion in the writable for select()"
        return self.producer_fifo or (not self.connected)
    ```